### PR TITLE
Tiny refactor and cleanup

### PR DESCRIPTION
**Pull request by fzyzcjy**
This is a very tiny code cleanup containing:

1. Refactor `update_sampling_params`. It originally uses a global variable and update states back and forth, and now uses immutable local variables.
2. Try to avoid mutate config in-place, which may cause a bit of confusion and bug in logic
3. Add `val_before_train: true` in the default config. This is because, otherwise users has to write `+trainer.val_before_train=...`, while all other configs has the style of `trainer.val_before_train=...`

~~Note: This PR starts from https://github.com/volcengine/verl/pull/74, thus the GitHub diff page will be clearer after #74 is merged.~~ (Ignore this line - that PR is already merged)

### CODE REVIEW

File: verl/trainer/config/ppo_megatron_trainer.yaml [hunk 1]
```diff
@@ -143,3 +143,4 @@ trainer:
   critic_warmup: 0
   default_hdfs_dir: ~/experiments/gsm8k/ppo/${trainer.experiment_name}
   default_local_dir: checkpoints/${trainer.project_name}/${trainer.experiment_name}
+  val_before_train: True
```
Comment (eric-haibin-lin) [inline verl/trainer/config/ppo_megatron_trainer.yaml:146]: Would you mind also updating the docs in https://github.com/volcengine/verl/blob/main/docs/examples/config.rst for val_before_train, and the forward batch size option added last time?

**Review by eric-haibin-lin — COMMENTED** [review]
thx for the contribution!

File: verl/trainer/config/ppo_megatron_trainer.yaml (hunk 1 above)
Comment (fzyzcjy) [reply verl/trainer/config/ppo_megatron_trainer.yaml:146]: Sure!

Comment (fzyzcjy) [reply verl/trainer/config/ppo_megatron_trainer.yaml:146]: Done

**Comment by PeterSH6** [comment]
Nice work! I'll review it later today.

**Comment by fzyzcjy** [comment]
> thx for the contribution!

You are welcome!

> Nice work!

Thanks!

> I'll review it later today.

Take your time :)

File: verl/trainer/fsdp_sft_trainer.py [hunk 2]
```diff
@@ -20,6 +20,8 @@
 
 import os
 
+from verl.utils.config import config_normalize_batch_size
```
Comment (PeterSH6) [inline verl/trainer/fsdp_sft_trainer.py:23]: Can you move this import above L35, which would be tidier

File: verl/utils/config.py [hunk 3]
```diff
@@ -14,10 +14,21 @@
...
+from omegaconf import DictConfig, open_dict
 
 
 def update_dict_with_config(dictionary: Dict, config: DictConfig):
     for key in dictionary:
         if hasattr(config, key):
             dictionary[key] = getattr(config, key)
+
+
+def config_normalize_batch_size(config, key: str, divider: int):
```
Comment (PeterSH6) [inline verl/utils/config.py:26]: I'm interested to know why in-place normalization might lead to confusion and how you feel when reading and using veRL's batch size configuration.

Additionally, I'd like to hear about any bugs you may have encountered while using in-place modification.

File: verl/workers/rollout/vllm_rollout/vllm_rollout.py [hunk 4]
```diff
@@ -118,26 +118,10 @@ def __init__(self, actor_module: nn.Module, config: DictConfig, tokenizer, model
...
         print(f"kwargs: {kwargs}")
-        self.sampling_params = SamplingParams(**kwargs)
+        self.partial_sampling_params = SamplingParams(**kwargs)
 
         self.pad_token_id = tokenizer.pad_token_id
 
-    @contextmanager
-    def update_sampling_params(self, **kwargs):
-        # update sampling params
-        old_sampling_params_args = {}
```
Comment (PeterSH6) [inline verl/workers/rollout/vllm_rollout/vllm_rollout.py:128]: Our original implementation also used a local variable while only maintaining one `SamplingParams`.

It seems that your refactor introduced deepcopy.
Could you clarify the benefit of your refactor more? Thanks!

**Review by PeterSH6 — CHANGES_REQUESTED** [review]

File: verl/workers/rollout/vllm_rollout/vllm_rollout.py (hunk 4 above)
Comment (fzyzcjy) [reply verl/workers/rollout/vllm_rollout/vllm_rollout.py:128]: Sure, I personally think the refactor is a little bit better, because we now avoid mutable global state, which seems to be generally not preferable in software engineering (e.g. a bit less readable and error-prone). The deepcopy is slower, but it seems the time cost is negligible, especially compared with the very heavy generate sequences.

Anyway, as the title suggests, this is just a tiny refactor, and feel free to revert it if you do not want :)

File: verl/utils/config.py (hunk 3 above)
Comment (fzyzcjy) [reply verl/utils/config.py:26]: As a simple example:

```
def __init__():
  f()
  ...
  normalize_batch_size()
  ...
  g()

def f():
  ...
  f2()

def f2():
  ...
  f3()

def f3():
  utilize config.batch_size

def g():
  utilize config.batch_size
```

Then, inside f3 we are using the un-normalized batch size, while in g we use the normalized one.
 
Surely this is just a made-up artificial case (and is very easy to realize something goes wrong), but in the real world I am worried whether, in the future, we will read batch size somewhere deeply nested, and accidentally it is *before* the normalize_batch_size, and then we are in trouble.

P.S. it is a bit about mutability in software engineering, which can introduce bugs when things unintentionally mutates. For example, in Rust, things are immutable by default.

Again, as is mentioned in the next comment, this is just a tiny refactor and feel free to accept or revert it :)

File: verl/trainer/fsdp_sft_trainer.py (hunk 2 above)
Comment (fzyzcjy) [reply verl/trainer/fsdp_sft_trainer.py:23]: (will do that in the next batch after discussions of other things)

File: verl/utils/config.py (hunk 3 above)
Comment (PeterSH6) [reply verl/utils/config.py:26]: Got it. I think your refactor regarding the in-place normalization is great.

However, I wonder how we deal with more complicated config mutate scenarios (such as $\times$ n in PR https://github.com/volcengine/verl/pull/80) if we follow the no-in-place mutate principle.

It would be better to keep only one additional config for each \*_batch_size config and we may need to make the normalize config function more general

Comment (fzyzcjy) [reply verl/utils/config.py:26]: > we may need to make the normalize config function more general

Totally agree! What about `config_normalize_batch_size(config, 'batch_size', self.config.rollout.n / self.device_mesh.shape[0])`, and at the same time change "divider" to "multiplier"?
(We may not face rounding issues since divider is always power of 2)

**Comment by eric-haibin-lin** [comment]
thanks for the previous efforts! Closing it due to multiple conflicts with the latest version. Feel free to open a new PR for new refactor/cleanup fixes. Thanks!!!

**Comment by fzyzcjy** [comment]
Sure, you are welcome! I will do that later when using verl again
